flink-ml-lib/src/main/java/org/apache/flink/ml/feature/standardscaler/OnlineStandardScalerModel.java

```diff
@@ -0,0 +1,298 @@
+
+        /** Model data for inference. */
+        private StandardScalerModelData modelData;
+
+        private DenseVector mean;
+
+        /** Inverse of standard deviation. */
+        private DenseVector scale;
+
+        private long modelVersion;
+
+        private long modelTimeStamp;
+
+        public PredictionOperator(
+                RowTypeInfo inputTypeInfo,
+                String inputCol,
+                boolean withMean,
+                boolean withStd,
+                long maxAllowedModelDelayMs,
+                String modelVersionCol) {
+            this.inputTypeInfo = inputTypeInfo;
+            this.inputCol = inputCol;
+            this.withMean = withMean;
+            this.withStd = withStd;
+            this.maxAllowedModelDelayMs = maxAllowedModelDelayMs;
+            this.modelVersionCol = modelVersionCol;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            bufferedPointsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<StreamRecord>(
+                                            "bufferedPoints",
+                                            new StreamElementSerializer(
+                                                    inputTypeInfo.createSerializer(
+                                                            getExecutionConfig()))));
+
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "modelData",
+                                            TypeInformation.of(StandardScalerModelData.class)));
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            if (modelData != null) {
+                modelDataState.clear();
+                modelDataState.add(modelData);
+            }
+        }
+
+        @Override
+        public void open() throws Exception {
+            super.open();
+            MetricGroup mlModelMetricGroup =
+                    getRuntimeContext()
+                            .getMetricGroup()
+                            .addGroup(MLMetrics.ML_GROUP)
+                            .addGroup(MLMetrics.ML_MODEL_GROUP);
+            mlModelMetricGroup.gauge(MLMetrics.TIMESTAMP, (Gauge<Long>) () -> modelTimeStamp);
+            mlModelMetricGroup.gauge(MLMetrics.VERSION, (Gauge<Long>) () -> modelVersion);
+        }
+
+        @Override
+        public void processElement1(StreamRecord<Row> dataPoint) throws Exception {
+            if (dataPoint.getTimestamp() <= modelTimeStamp + maxAllowedModelDelayMs
```

Comment:
According to the maxAllowedModelDelayMs doc, this parameter should only be used if the input contains event time. I suppose there should be logic that differentiates between event time and system time. Can you explain where is this logic?

Reply:
Good catch, it is indeed a bug. I have fixed the training logic as follows:
- When the window strategy is based on event-time, the timestamp of model data is the maximum timestamp of the window that generates this model.
- Otherwise, the timestamp of the model data is `LONG.MAX_VALUE`.

Moreover, to avoid the possible overflow, the `if condition` here is updated as `if (dataPoint.getTimestamp() - maxAllowedModelDelayMs <= modelTimeStamp`.

Reply:
Thanks for the fix. LGTM.